### PR TITLE
doc(davide): add GTrXL doc

### PR DESCRIPTION
**Pull request by davide97l**

### CODE REVIEW

File: source/hands_on/gtrxl.rst [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+2. GTrXL only supports **sequential** observations.
+
+3. GTrXL is based on **Transformer-XL** with **Gating connections**.
+
+4. The DI-engine implementation of GTrXL is based on the R2D2 algorithm. In the original paper, it is based on the algorithm V-MPO.
+
+Key Equations or Key Graphs
+---------------------------
+**Transformer-XL**: is a Transformer architecture that introduces the notion of recurrence to the deep self-attention network.
```
Comment (puyuan1996) [inline source/hands_on/gtrxl.rst:25]: To address the context fragmentation problem, Transformer-XL introduces the notion of recurrence to the deep self-attention network.

File: source/hands_on/gtrxl.rst [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+Instead of computing the hidden states from scratch for each new segment, Transformer-XL reuses the hidden states obtained in previous segments.
+The reused hidden states serve as memory for the current segment, which builds up a recurrent connection between the segments.
+As a result, modeling very long-term dependency becomes possible because information can be propagated through the recurrent connections.
+As an additional contribution, the Transformer-XL uses a new relative positional encoding formulation that generalizes to attention lengths longer than the one observed during training.
```
Comment (puyuan1996) [inline source/hands_on/gtrxl.rst:29]: In order to enable state reuse without causing temporal confusion, Transformer-XL proposes a new relative positional encoding formulation that generalizes to attention lengths longer than the one observed during training.

File: source/hands_on/gtrxl.rst [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+.. image:: images/transformerXL_train_eval.png
+   :align: center
+
+**Identity map reordering**: move the layer normalization to the input stream of the submodules.
```
Comment (puyuan1996) [inline source/hands_on/gtrxl.rst:34]: Identity Map Reordering

File: source/hands_on/gtrxl.rst [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+With identity map reordering the forward pass of the model can be computed as:
+
+.. image:: images/identity_map_reordering.png
+   :align: center
+
+.. image:: images/gtrxl.png
+   :align: center
+   :height: 300
+
+**Gating layers**: replace the residual connections with gating layers. Among several studied gating functions, Gated Recurrent Unit (GRU) is the one that performs the best.
```
Comment (puyuan1996) [inline source/hands_on/gtrxl.rst:49]: Gating Layers

File: source/hands_on/gtrxl.rst [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+As an additional contribution, the Transformer-XL uses a new relative positional encoding formulation that generalizes to attention lengths longer than the one observed during training.
+
+.. image:: images/transformerXL_train_eval.png
+   :align: center
+
+**Identity map reordering**: move the layer normalization to the input stream of the submodules.
+A key benefit to this reordering is that it now enables an identity map from the input of the transformer at the first layer to the output of the transformer after the last layer.
+This is in contrast to the canonical transformer, where there are a series of layer normalization operations that non-linearly transform the state encoding.
+One hypothesis as to why the Identity Map Reordering improves results is as follows: assuming that the submodules at initialization produce values that are in expectation near
+zero, the state encoding is passed un-transformed to the policy and value heads, enabling the agent to learn a Markovian policy at the start of training.
```
Comment (puyuan1996) [inline source/hands_on/gtrxl.rst:38]: Why when the submodules at initialization produce values that are in expectation near
zero, then the state encoding is passed un-transformed to the policy and value heads and why in this case, the agent learn a Markovian policy? I don't quite understand. Could you explain more thoroughly?

File: source/hands_on/gtrxl.rst [hunk 6]
```diff
@@ -0,0 +1,139 @@
+GTrXL
+^^^^^^^
+
+Overview
+---------
+Gated Transformer-XL, or GTrXL, first proposed in `Stabilizing Transformers for Reinforcement Learning <https://arxiv.org/pdf/1910.06764.pdf>`_,
+is a novel framework for reinforcement learning based on the Transformer-XL architecture.
+It introduces several architectural modifications that improve the stability and learning speed of Transformer including:
```
[the remark this answers is not in the document]
Comment (davide97l) [reply source/hands_on/gtrxl.rst:8]: changed

File: source/hands_on/gtrxl.rst (hunk 1 above)
Comment (davide97l) [reply source/hands_on/gtrxl.rst:25]: added

File: source/hands_on/gtrxl.rst (hunk 2 above)
Comment (davide97l) [reply source/hands_on/gtrxl.rst:29]: changed

File: source/hands_on/gtrxl.rst (hunk 5 above)
Comment (davide97l) [reply source/hands_on/gtrxl.rst:38]: image
You have to look at this figure to understand: if the output of the attention block is near 0, the input of the GRU is close to the input observation. Attention block near 0 means that the effect of past behaviors is also negligible.
image
The content between the brackets will make it more clear, i will explain better the doc

**Comment by davide97l** [comment]
Benchmark will be added once GTrXL and R2D2 optimization is completed

**Pull request merged**